src/sentry/region_to_control/producer.py

```diff
@@ -1,91 +1,178 @@
+        it, caching it for future calls.  Installs an exit handler to close the worker thread processes.
+        """
+        if self._publisher is None:
+            config = settings.KAFKA_TOPICS.get(settings.KAFKA_REGION_TO_CONTROL)
+            self._publisher = KafkaProducer(
+                kafka_config.get_kafka_producer_cluster_options(config["cluster"])
+            )
+
+            @atexit.register
+            def exit_handler():
+                self.close()
+
+        return self._publisher
+
+
+# Test default mock.
+class MockRegionToControlMessageService(RegionToControlMessagingService):
+    def __init__(self):
+        from unittest.mock import Mock
+
+        self.mock = Mock(spec_set=self)
+
+    def write_region_to_control_message(self, payload: Any, sync: bool):
+        self.mock.write_region_to_control_message(payload, sync)
 
-def produce_audit_log_entry(entry: AuditLogEntry):
-    if _should_produce_to_kafka():
-        write_region_to_control_message(
+    def close(self):
+        self.mock.reset_mock()
+
+
+# This service should only really exist in the region mode for now, but the delegator object is used for its
+# convenience methods.
+region_to_control_message_service: RegionToControlMessagingService = silo_mode_delegation(
+    {SiloMode.REGION: KafkaBackedRegionToControlMessageService}
+)
+
+
+class AuditLogEntryService(InterfaceWithLifecycle):
+    @abstractmethod
+    def produce_audit_log_entry(self, entry: AuditLogEntry):
+        pass
+
+
+class RegionToControlAuditLogEntryService(AuditLogEntryService):
+    region_to_control_messaging_service: RegionToControlMessagingService
+
+    def __init__(
+        self,
+        region_to_control_messaging_service: RegionToControlMessagingService = region_to_control_message_service,
+    ):
+        self.region_to_control_messaging_service = region_to_control_messaging_service
+
+    def produce_audit_log_entry(self, entry: AuditLogEntry):
+        self.region_to_control_messaging_service.write_region_to_control_message(
             dataclasses.asdict(RegionToControlMessage(audit_log_event=entry.as_kafka_event())),
             sync=True,
         )
-    else:
+
+    def close(self):
+        pass
+
+
+class DatabaseBackedAuditLogEntryService(AuditLogEntryService):
+    def produce_audit_log_entry(self, entry: AuditLogEntry):
         entry.save()
 
+    def close(self):
+        pass
 
-_publisher: Optional[KafkaProducer] = None
-_empty_headers = list()
 
+audit_log_entry_service: AuditLogEntryService = silo_mode_delegation(
+    {
+        SiloMode.MONOLITH: DatabaseBackedAuditLogEntryService,
+        SiloMode.CONTROL: DatabaseBackedAuditLogEntryService,
+        SiloMode.REGION: RegionToControlAuditLogEntryService,
+    }
+)
+
+
+class UserIpService(InterfaceWithLifecycle):
+    @abstractmethod
+    def produce_user_ip(self, event: UserIpEvent):
```

Comment:
Could call this `record_event` :shrug:

Reply:
I tend to prefer slightly?  more descriptive contextual names because it makes grepping easier.  And right now, my current refactor heavily benefits from method names that are easy to grep.